lib/auth/auth.js

```diff
@@ -26,120 +30,153 @@ auth.setAuthHandler = handler => {
+        }
+        log.trace('identified auth method', { version, authMethod: method });
+        return checkFunctions[version][method](request, log, data, awsService);
+    }
+
+    // no auth info identified
+    log.debug('assuming public user');
+    return { err: null, params: publicUserInfo };
+}
+
+function doAuth(request, log, cb, awsService, requestContext) {
+    const res = extractParams(request, log, awsService, request.query);
     if (res.err) {
         return cb(res.err);
-    } else if (res.version === 2) {
-        return vault.authenticateV2Request(res.data.accessKey,
-            res.data.signatureFromRequest,
-            res.data.stringToSign, res.data.algo, requestContext, log,
-            (err, authInfo) => {
-                if (err) {
-                    return cb(err);
-                }
-                return cb(null, authInfo);
-            });
-    } else if (res.version === 4) {
-        res.data.log = log;
-        return vault.authenticateV4Request(res.data, requestContext,
-            cb, awsService);
-    } else if (res.data instanceof AuthInfo) {
-        return cb(null, res.data);
+    } else if (res.params instanceof AuthInfo) {
+        return cb(null, res.params);
     }
+
+    requestContext.setAuthType(res.params.data.authType);
+    requestContext.setSignatureVersion(res.params.data.signatureVersion);
+    requestContext.setSignatureAge(res.params.data.signatureAge);
+
+    // Corner cases managed, we're left with normal auth
+    res.params.log = log;
+    if (res.params.version === 2) {
+        return vault.authenticateV2Request(res.params, requestContext, cb);
+    } else if (res.params.version === 4) {
+        return vault.authenticateV4Request(res.params, requestContext, cb,
+                                           awsService);
+    }
+
     log.error('authentication method not found', {
         method: 'Arsenal.auth.doAuth',
     });
     return cb(errors.InternalError);
-};
+}
 
-auth.generateV4Headers =
-    (request, data, accessKey, secretKeyValue, awsService) => {
-        Object.assign(request, { headers: {} });
-        const amzDate = convertUTCtoISO8601(Date.now());
-        // get date without time
-        const scopeDate = amzDate.slice(0, amzDate.indexOf('T'));
-        const signedHeaders = 'host;x-amz-date;x-amz-content-sha256';
-        const region = 'us-east-1';
-        const service = awsService || 'iam';
-        const credentialScope =
-            `${scopeDate}/${region}/${service}/aws4_request`;
-        const timestamp = amzDate;
-        const algorithm = 'AWS4-HMAC-SHA256';
+function generateV4Headers(request, data, accessKey, secretKeyValue,
+                           awsService) {
+    Object.assign(request, { headers: {} });
+    const amzDate = convertUTCtoISO8601(Date.now());
+    // get date without time
+    const scopeDate = amzDate.slice(0, amzDate.indexOf('T'));
+    const signedHeaders = 'host;x-amz-date;x-amz-content-sha256';
```

Comment:
I didn't really refine the generateV4Headers function, as you can see from the diff.
It wasn't really a priority for this first step of the refactor, but then any suggestion brought to issue #102 should be taken into account for the next steps of this work.
Thanks for the input :)